core : Float constructor allows to set binary or decimal precision

ArifAhmed1995 opened the pull request:
Either binary or decimal precision is supplied to the constructor of Float.
If neither is supplied the default precision of 15 decimal digits is used.
If both are supplied an error is thrown , even if decimal and binary precision
match.

Ref : #12215 
<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below.>
Fixes #12215

ArifAhmed1995 committed:
'Printing' changes and fix rebase of previous commit

ArifAhmed1995 committed:
core: Float constructor allows to set binary or decimal precision.

Either binary or decimal precision is supplied to the constructor of Float.
If neither is supplied the default precision of 15 decimal digits is used.
If both are supplied an error is thrown , even if decimal and binary precision
match.

asmeurer commented on sympy/core/numbers.py:
```diff
@@ -768,7 +768,11 @@ class Float(Number):
-    def __new__(cls, num, prec=None):
+    def __new__(cls, num, dps=None, prec=None):
+        if dps and prec is not None:
+            raise ValueError('Both decimal and binary precision supplied.'
```
Need a space after the `.`

asmeurer commented on sympy/core/numbers.py:
```diff
@@ -768,7 +768,11 @@ class Float(Number):
 
-    def __new__(cls, num, prec=None):
+    def __new__(cls, num, dps=None, prec=None):
+        if dps and prec is not None:
```
Just `if dps and prec` would also work, and would read clearer.

asmeurer replied:
And also in the code below, you can just check `prec` instead of `prec is None`, because it will evaluate to True if it is a nonzero integer (and 0 isn't a valid precision).

asmeurer replied:
Oh I didn't know about the `prec=''` behavior. Never mind then.

asmeurer commented:
This looks good. Can you add some tests for creating floats with precisions that aren't possible with dps?

cbm755 commented:
Cool!

A minor nitpick:
> If neither is supplied the default precision of 15 decimal digits is used.
This is not correct; the behaviour is more complicated than that.  More accurate statement is "... default precision of at least 15 decimal digits is used".

cbm755 commented on sympy/printing/tests/test_repr.py:
```diff
@@ -101,12 +101,13 @@ def test_Rational():
+    sT(Float('1.234567890123456789', prec=66),
+       "Float('1.23456789012345678901321800735590983322254032827913761138916015625', prec=66)")
+    sT(Float('0.60038617995049726', 15),
+       "Float('0.6003861799504972562857574303052388131618499755859375', prec=53)")
```
This last test doesn't look right.  `prec=53` is roughly `dps = 15` right?  So why is it producing 53 decimal digits?  I think the last line should be
````
sT(Float('0.60038617995049726', 15), Float('0.60038617995049726', prec=53))
````
(or something very close to that).

ArifAhmed1995 replied:
Yes , the `_print_Float()` in `repr.py` uses the `to_str` function of mpmath. The `to_str` function uses decimal precision to yield the string representation of a mpmath float. 

I'll make the required changes.

cbm755 reviewed (changes requested):
As I said, I like this idea, here are a few changes.

asmeurer commented:
@cbm755 you are right. I didn't realize that `srepr` printed the keyword argument name. So this is a major compatibility break. Let's call the argument `_prec` and add a note that it should be changed in a future version.

asmeurer commented:
(`_prec` seems the best name to me because that's also the name of the attribute)

asmeurer commented:
OTOH `_prec` suggests it is a private API, which it isn't. So maybe `precision` is better. What do you guys think?

cbm755 commented:
`prec_`?   `precision` was the best I could come up with (see my comments at the original issue for more detail).

ArifAhmed1995 commented on sympy/core/numbers.py:
```diff
@@ -784,11 +788,14 @@ def __new__(cls, num, prec=None):
+                if dps is None:
+                    prec = num.context.prec
+            else:
+                prec = num.context.prec
```
Ah, yes. I was just beginning to write some tests and I spotted this discrepancy.Thanks for pointing it out.

ArifAhmed1995 committed:
Implement requested changes:

1.Add Deprecation Warning for use of 'prec=XX'.
2.Add binary precision tests.
3.PEP8 fixes.

cbm755 commented on sympy/core/numbers.py:
```diff
@@ -803,10 +821,11 @@ def __new__(cls, num, prec=None):
             if not isinstance(num, string_types):
                 raise ValueError('The null string can only be used when '
-                'the number to Float is passed as a string or an integer.')
+                'the number to Float is passed as a string or an integer. ')
```
Remove spurious trailing space

cbm755 commented on sympy/core/numbers.py:
```diff
@@ -803,10 +821,11 @@ def __new__(cls, num, prec=None):
                     dps = max(15, dps)
-        elif prec == '':
+                    precision = mlib.libmpf.dps_to_prec(dps)
+        elif precision == '' and dps == None or precision == None and dps == '':
```
'is None' rather than =

cbm755 commented on sympy/printing/tests/test_repr.py:
```diff
@@ -101,12 +101,11 @@ def test_Rational():
 
 
 def test_Float():
-    sT(Float('1.23', prec=3), "Float('1.22998', prec=3)")
```
I think these tests should stay almost "as is", just with dps=.  Then you can add your new tests as you've done here.  Logic here is nothing wrong with the old tests so leave then alone as Mich as possible .

ArifAhmed1995 replied:
In my view, it's better to always display in terms of binary precision to maintain the characteristics of the original Float. For example, we should print `Float('XXXX', precision = YYY)` if the Float was originally created with binary precision.

According to me(as per Sympy 1.0) , if I create : `Float('1.23', prec=3)` , I would expect the same to be printed out , because I explicitly mentioned `3` decimal digits.I don't see why it should be printed out as something different.But then again ,I may be incorrect in this.

I think that the value is stored correctly because both `a._mpf_` and `eval(srepr(a))._mpf_` are same.

Do you think it would be better if `srepr` prints with `dps=YY` when original Float is created with decimal precision and `precision=YY` for all other cases?

ArifAhmed1995 replied:
@asmeurer Would like to hear your thoughts on this `srepr` issue.

cbm755 replied:
> it's better to always display in terms of binary precision 

Yes I agree.

My concern is that srepr printing has changed: before it was `Float('1.22998', prec=3)` now it is `Float('1.23', precision=13)`.  I agree that 13 is the right thing to print.  But why has displayed digits changed?  Its not necessarily bad; just not yet explained.

cbm755 replied:
> Do you think it would be better if srepr prints with dps=YY when original Float is created with decimal precision and precision=YY for all other cases?

I don't think so; I like how you have it.

ArifAhmed1995 replied:
It can be explained because an extra padding of 3 decimal digits is provided to make sure that no precision in lost between translations from decimal to binary(at least I think that's the case). Since now the translation from decimal to binary precision is taken care of by the `dps_to_prec` call in the `Float` constructor, there is no need for it.
```
def repr_dps(n):
    """Return the number of decimal digits required to represent
    a number with n-bit precision so that it can be uniquely
    reconstructed from the representation."""
    dps = prec_to_dps(n)
    if dps == 15:
        return 17
    return dps + 3
```

cbm755 replied:
>  no precision in lost between translations from decimal to binary

Right, so that confirms why the extra 3 digits were there before.  Thanks for this.

> Since now the translation from decimal to binary precision is taken care

But I'm not sure the conclusion follows (that the extra digits are now unnecessary).  although I've yet to find a counterexample.

ArifAhmed1995 replied:
`srepr` is used just to yield the string representation of the given Float, not for any further computation. I don't see why one would want to see extra decimal digits as compared to imput.

cbm755 replied:
You said it yourself:

> to make sure that no precision in lost between translations from decimal to binary

(This means between decimal and binary *representations* of the number, not between `dps` and `precision`.)

> not for any further computation.

No so, `eval(srepr(x))` should give exact same object whenever possible.

cbm755 replied:
To elaborate: one might conclude that srepr of a float should prints its `_mpf`_`; but this is not so readable.  We want a similar guarantee of `eval(srepr(...))` being identity map.

ArifAhmed1995 replied:
Ah yes , I now remember that the `sT()` function in `test_repr` does an evaluation on `srepr`.

cbm755 reviewed (changes requested):
Looking good!

ArifAhmed1995 committed:
Add extra tests, PEP8 fixes

cbm755 commented:
Both of these should be explicitly disallowed
````
Float("1.23", dps="", precision=7)
Float("1.23", dps=5, precision="")
````

cbm755 commented on sympy/core/numbers.py:
```diff
@@ -768,7 +771,20 @@ class Float(Number):
 
-    def __new__(cls, num, prec=None):
+    def __new__(cls, num, dps=None, prec=None, precision=None):
+        if dps and precision or dps and prec or prec and precision is not None:
```
maybe you should move the `prec` deprecation stuff before this test so there is not `prec` in this if?

cbm755 commented:
We need some `raises(ValueError, lambda: Float("1.1", dps=3, precision=100)`, and also for the two cases I mentioned above.

cbm755 commented on sympy/core/numbers.py:
```diff
@@ -818,20 +837,21 @@ def __new__(cls, num, prec=None):
-            dps = prec
 
-        prec = mlib.libmpf.dps_to_prec(dps)
+        if precision is None:
```
How about a comment here to help our future selves understand the logic?

````
# after the above, dps is definitely set and precision might be too.
assert dps is not None
if prec is None:
     ...
````

ArifAhmed1995 committed:
Add extra tests and documentation

ArifAhmed1995 commented on sympy/core/tests/test_numbers.py:
```diff
@@ -487,14 +488,30 @@ def teq(a):
+    # Decimal value 0.1 cannot be expressed precisely as a base 2 fraction
+    Float(S(1)/10, precision=53) == S(0.1)
+    Float(S(1)/10, precision=53) == S(0.1)
+    assert not Float(S(1)/10, precision=53) == Float(S(1)/10, precision=54)
```
@asmeurer Is this the kind of test you talked about earlier ?

ArifAhmed1995 commented:
@cbm755 I've made the required changes.Can you please review again ?

ArifAhmed1995 committed:
Rectify a test and grammar

cbm755 commented on sympy/core/tests/test_numbers.py:
```diff
@@ -487,14 +488,30 @@ def teq(a):
+    # Decimal value 0.1 cannot be expressed precisely as a base 2 fraction
+    Float(S(1)/10, precision=53) == S(0.1)
+    Float(S(1)/10, precision=54) == S(0.1)
+    assert not Float(S(1)/10, precision=53) == Float(S(1)/10, precision=54)
```
I think explicitly testing for this bullshit non-associate structural equality is a bad idea.  i.e., this is `a == b` and `c == b` but `a != c`.  This isn't your fault, but let's not build it into the tests.

Would this be a better test?
````

In [34]: a = Float(S(1)/10, dps=15)

In [35]: b = Float(S(1)/10, dps=16)

In [36]: p = Float(S(1)/10, precision=53)

In [37]: q = Float(S(1)/10, precision=54)

In [38]: a._mpf_ == p._mpf_
Out[38]: True

In [39]: a._mpf_ == q._mpf_
Out[39]: False

In [40]: b._mpf_ == q._mpf_
Out[40]: False
````
This shows that:
1.  `a` and `p` are the same
2.  But there are some bit precisions with no corresponding `dps`.  (`q` is "between" `a` and `b`).

ArifAhmed1995 replied:
Yes, this is a better test

ArifAhmed1995 committed:
Improve test for binary precision

cbm755 commented on sympy/printing/repr.py:
```diff
@@ -135,9 +135,8 @@ def _print_Fraction(self, expr):
-        dps = prec_to_dps(expr._prec)
-        r = mlib.to_str(expr._mpf_, repr_dps(expr._prec))
-        return "%s('%s', prec=%i)" % (expr.__class__.__name__, r, dps)
+        r = mlib.to_str(expr._mpf_, prec_to_dps(expr._prec))
```
Let's leave this as `repr_dps`.  You can always do another PR to change/discuss this, but it seems a shame to hold up this good `dps` change.

Its easier to get PRs that do one thing passed.  This current PR does two things: the dps/precision change and a (likely unrelated) srepr printing change.

ArifAhmed1995 replied:
Cool. I've done the changes.

ArifAhmed1995 replied:
@cbm755 Can this be merged now ? Another PR depends on the changes made here #11666

cbm755 commented on sympy/core/tests/test_numbers.py:
```diff
@@ -1,4 +1,5 @@
 import decimal
+
```
extra blank line has snuck in

ArifAhmed1995 committed:
Revert to repr_dps

cbm755 commented:
> Can this be merged now ? Another PR(#11666) depends on the changes made here

I think so, you've addresses my concerns anyway and I think @asmeurer's comments as well.

I will merge in 24 hours or so.

While we wait, can you also draft some text [for the release notes](https://github.com/sympy/sympy/wiki/Release-Notes-for-1.1)?  (to note the API change.)

ArifAhmed1995 commented:
@cbm755 I've added the deprecation change to the release notes.

asmeurer commented:
A deprecation issue was never opened for this. It's now https://github.com/sympy/sympy/issues/12820. Pull request adding it to the code here https://github.com/sympy/sympy/pull/12821.
